chardet/utf1632prober.py

```diff
@@ -0,0 +1,137 @@
+######################## BEGIN LICENSE BLOCK ########################
+# The Original Code is mozilla.org code.
+#
+# The Initial Developer of the Original Code is
+# Netscape Communications Corporation.
+# Portions created by the Initial Developer are Copyright (C) 1998
+# the Initial Developer. All Rights Reserved.
+#
+# Contributor(s):
+#   Jason Zavaglia
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+# 02110-1301  USA
+######################### END LICENSE BLOCK #########################
+from chardet.enums import ProbingState
+from .charsetprober import CharSetProber
+
+
+# This class simply looks for occurrences of zero bytes, and infers
+# whether the file is UTF16 or UTF32 (low-endian or big-endian)
+# For instance, files looking like ( \0 \0 \0 [nonzero] )+
+# appear to be UTF32LE.  Files looking like ( \0 [notzero] )+ appear
+# may be guessed to be UTF16LE.
+
+class UTF1632Prober(CharSetProber):
```

Comment:
The way we would usually handle this sort of multi-byte encoding is with a [`CodingStateMachine`](https://github.com/chardet/chardet/blob/master/chardet/codingstatemachine.py#L33) and [`CharDistributionAnalysis`](https://github.com/chardet/chardet/blob/master/chardet/chardistribution.py#L40)-based implementation. You can see the prober we have for UTF-8 (which is obviously single-byte) [here](https://github.com/chardet/chardet/blob/master/chardet/utf8prober.py), which only uses the `CodingStateMachien`, but the approach is close to what we would need to do for UTF-16 and UTF-32. In fact, through some crazy bit of coincidence, we already have [unused state machines for UTF-16LE and UTF-16BE in our code](https://github.com/chardet/chardet/blob/master/chardet/mbcssm.py#L390-L498) that I've never noticed until now. It appears to have been there since version 1.0, but never used.

Anyway, I know it would be a bit of work to switch this over to using `CodingStateMachine` like the UTF-8 prober, so I'm not going to require that to merge this, but if you don't do it, I'll probably end up switching this over to that eventually.